src/main/java/com/arpnetworking/kairosdb/HistogramDataPointV2Impl.java

```diff
@@ -0,0 +1,249 @@
+                .add("max", _max)
+                .add("mean", _mean)
+                .add("sum", _sum)
+                .add("timestamp", m_timestamp)
+                .toString();
+    }
+
+    @Override
+    public void writeValueToBuffer(final DataOutput buffer) throws IOException {
+        final DataPointV2.DataPoint.Builder builder = DataPointV2.DataPoint.newBuilder();
+
+        for (final Map.Entry<Double, Integer> entry : _map.entrySet()) {
+            builder.putHistogram(pack(entry.getKey()), entry.getValue());
+        }
+
+        builder.setMax(_max);
+        builder.setMin(_min);
+        builder.setMean(_mean);
+        builder.setSum(_sum);
+        builder.setPrecision(_precision);
+
+
+        final byte[] bytes = builder.build().toByteArray();
+        buffer.writeInt(bytes.length);
+        buffer.write(bytes);
+    }
+
+    @Override
+    public void writeValueToJson(final JSONWriter writer) throws JSONException {
+        writer.object().key("bins");
+        writer.object();
+        for (final Map.Entry<Double, Integer> entry : _map.entrySet()) {
+            writer.key(entry.getKey().toString()).value(entry.getValue());
+        }
+        writer.endObject();
+        writer.key("min").value(_min);
+        writer.key("max").value(_max);
+        writer.key("mean").value(_mean);
+        writer.key("sum").value(_sum);
+        writer.key("precision").value(_precision);
+        writer.endObject();
+    }
+
+    @Override
+    public String getApiDataType() {
+        return API_TYPE;
+    }
+
+    @Override
+    public String getDataStoreDataType() {
+        return HistogramDataPointV2Factory.DST;
+    }
+
+    @Override
+    public boolean isLong() {
+        return false;
+    }
+
+    @Override
+    public long getLongValue() {
+        return 0;
+    }
+
+    @Override
+    public boolean isDouble() {
+        return false;
+    }
+
+    @Override
+    public double getDoubleValue() {
+        return 0;
+    }
+
+    @Override
+    public long getSampleCount() {
+        return _countSupplier.get();
+    }
+
+    private long computeSampleCount() {
+        long count = 0;
+        for (final Integer binSamples : _map.values()) {
+            count += binSamples;
+        }
+        return count;
+    }
+
+    @Override
+    public long getOriginalCount() {
+        return _originalCountSupplier.get();
+    }
+
+
+    @Override
+    public int getPrecision() {
+        return _precision;
+    }
+
+    @Override
+    public double getSum() {
+        return _sum;
+    }
+
+    @Override
+    public double getMin() {
+        return _min;
+    }
+
+    @Override
+    public double getMax() {
+        return _max;
+    }
+
+    @Override
+    public TreeMap<Double, Integer> getMap() {
+        return _map;
+    }
+
+    long pack(final double unpacked) {
+        final long truncated = Double.doubleToRawLongBits(unpacked) & _mask;
+        final long shifted = truncated >> (52 - _precision);
```

Comment:
Where does the `52` come from?

Reply:
The logic for this one is a bit complicated. It has to do with trying to shift out as many 0's as we could. And then also re-arrange the bits such that we get the smallest numbers (probabilistic). The idea there is that smaller numbers are smaller (in bits) in the variable encoding that we use (protobuf).